packages/pigeon/lib/ast.dart

```diff
@@ -82,35 +74,101 @@ class Api extends Node {
   }
 }
 
+/// A parameter to a generic entity.  For example, "String" to "List<String>".
+class TypeArgument {
+  /// Constructor for [TypeArgument].
+  TypeArgument({
+    required this.dataType,
+    required this.isNullable,
+    this.typeArguments,
+  });
+
+  /// A string representation of the base datatype.
+  final String dataType;
+
+  /// The type arguments to this [TypeArgument].
+  final List<TypeArgument>? typeArguments;
+
+  /// True if the type is nullable.
+  final bool isNullable;
+
+  @override
+  String toString() {
+    return '(TypeArgument dataType:$dataType isNullable:$isNullable typeArguments:$typeArguments)';
+  }
+}
+
+/// An entity that represents a typed concept, like a [TypeArgument] or [Field].
+abstract class TypedEntity {
+  /// The data-type of the entity (ex 'String' or 'int').
```

Comment:
I find this description and the examples (and even the name of the getter) misleading. If something is a `List<String>?` its type is that, not `List`. Is there a standard name for this component of the type in Dart? If not, maybe 'baseDataType'?

At the very least, the example should have something like that `List<String>?`->`List` example, rather than only having examples where the two other fields don't apply.